server/command.go

```diff
@@ -0,0 +1,409 @@
+		p.postCommandResponse(context, noChannelSubscription)
+		return &model.CommandResponse{}
+	}
+	list := serializer.FormattedSubscriptionList(channelSubscriptions)
+	p.postCommandResponse(context, list)
+	return &model.CommandResponse{}
+}
+
+func confluenceHelpCommand(p *Plugin, context *model.CommandArgs, args ...string) *model.CommandResponse {
+	helpText := getFullHelpText(p, context, args...)
+
+	p.postCommandResponse(context, helpText)
+	return &model.CommandResponse{}
+}
+
+func getFullHelpText(p *Plugin, context *model.CommandArgs, args ...string) string {
+	helpText := commonHelpText
+	if utils.IsSystemAdmin(context.UserId) {
+		helpText += sysAdminHelpText
+	}
+	return helpText
+}
+
+func executeConnect(p *Plugin, context *model.CommandArgs, args ...string) *model.CommandResponse {
+	instances, err := p.instanceStore.LoadInstances()
+	if err != nil {
+		return p.responsef(context, "Failed to load instances. Error: %v.", err)
+	}
+	confluenceURL := ""
+	if len(args) > 0 {
+		confluenceURL = args[0]
+	}
+	instance := instances.getByAlias(confluenceURL)
+	if instance != nil {
+		confluenceURL = instance.InstanceID.String()
+	}
+	isAdmin := len(args) > 1 && strings.EqualFold(args[1], AdminMattermostUserID)
+	info, err := p.GetUserInfo(types.ID(context.UserId), nil)
+	if err != nil {
+		return p.responsef(context, "Failed to connect. Error: %v", err)
+	}
+	if info.Instances.IsEmpty() {
+		return p.responsef(context,
+			"No Confluence instances have been installed. Please contact the system administrator.")
+	}
+	if confluenceURL == "" {
+		if info.connectable.Len() == 1 {
+			confluenceURL = info.connectable.IDs()[0].String()
+		}
+	}
+	instanceID := types.ID(confluenceURL)
+	if info.connectable.IsEmpty() {
+		return p.responsef(context,
+			"You already have connected all available Confluence accounts. Please use `/confluence disconnect --instance=%s` to disconnect.",
+			instanceID)
+	}
+	if !info.connectable.Contains(instanceID) {
+		return p.responsef(context,
+			"Confluence instance %s is not installed, please contact the system administrator.",
+			instanceID)
+	}
+	conn, err := p.userStore.LoadConnection(instanceID, types.ID(context.UserId))
+	if err == nil && len(conn.ConfluenceAccountID()) != 0 {
+		return p.responsef(context,
+			"You already have a Confluence account linked to your Mattermost account from %s. Please use `/confluence disconnect --instance=%s` to disconnect.",
+			instanceID, instanceID)
+	}
+	if _, ok := p.getConfig().ParsedConfluenceConfig[confluenceURL]; !ok {
+		return p.responsef(context, configNotFoundError, instanceID, instanceID)
+	}
+
+	link := instancePathWithAdmin(routeUserConnect, instanceID, isAdmin)
+	return p.responsef(context, "[Click here to link your Confluence account](%s%s)", p.GetPluginURL(), link)
+}
+
+func executeDisconnect(p *Plugin, commArgs *model.CommandArgs, args ...string) *model.CommandResponse {
+	if len(args) > 1 {
+		return p.help(commArgs)
+	}
+	confluenceURL := ""
```

Comment:
not sure about the confluence URL logic. If there is no URL, shouldn't we get out of here as soon as possible? we would remove some of the problems that may arise in lines:

- 365 `instance := instances.getByAlias(confluenceURL)`
- 369 `disconnected, err := p.DisconnectUser(confluenceURL, types.ID(commArgs.UserId))`
- 372 (this check wouldn't be needed)

Is there a case where the ConfluenceURL would be empty?

Reply:
actually, this is for /confluence connect <URL>. but here this URL is optional in the case if there is only 1 instance installed then it will automatically take the URL for that instance and if multiple instances are installed then the user has to specify the URL. Here are all the edge cases are covered